cassiopeia/solver/ilp_solver_utilities.pyx

```diff
@@ -93,13 +237,90 @@ def infer_layer_of_potential_graph(
         for k in range(len(top_ancestors)):
             if distance_to_ancestors[k] == min_distance_to_ancestor:
 
-                parent = top_ancestors[k][:dim]
+                # The first (2*dim-1) characters correspond to the parent
+                parent = "|".join(top_ancestors[k].split("|")[:dim])
                 edge2 = top_ancestors[k]
-                edge1 = np.concatenate((parent, sample1))
+                edge1 = parent + "|" + sample1
 
                 new_edges.append(edge1)
                 new_edges.append(edge2)
 
                 layer.append(parent)
 
-    return np.array(layer), np.array(new_edges)
\ No newline at end of file
+    return np.array(layer), np.array(new_edges)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def get_lca_characters_cython(
+    arr1: np.array(),
+    arr2: np.array(),
+    n_char: int,
+    missing_state_indicator: str,
+) -> str:
+    """Cython-optimized LCA inference.
+
+    A utility function for the inference of the latest common ancestor between
+    two samples. Here, both samples are lists of strings where each value
+    corresponds to a state. The procedure infers an ancestral state according to
+    Camin-Sokal parsimony and returns an ancestor character string where each
+    state is delimited by `|`.
+
+    Args:
+        arr1: Character array of sample 1
+        arr2: Character array of sample 2
+        n_char: Number of characters in array
+        missing_state_indicator: Indicator for missing data.
+
+    Returns:
+        The character string of the inferred ancestor, where each character is
+            delimited by a `|`.
+    """
+
+    cdef list ancestor = ["0"]*n_char
+    cdef int i
+
+    for i in range(n_char):
+
+        if arr1[i] == arr2[i]:
+            ancestor[i] = arr1[i]
+
+        else:
+            if arr1[i] == missing_state_indicator:
+                ancestor[i] = arr2[i]
+            elif arr2[i] == missing_state_indicator:
+                ancestor[i] = arr1[i]
+
+    return "|".join(ancestor)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def simple_hamming_distance_cython(
```

Comment:
Hey Matt so I'm not super familiar with cython, do we need both this cythonized version and the non-cythonized version of this function because we have to call something like this in cythonized functions and non-cythonized functions? We can't collapse these?

Reply:
Two reason -- the first is that the cythonized version works a lot faster here, and it operates on strings not integers.

Also, I'm intentionally ignoring missing data here, which is how the original codebase worked (I think it's okay for inferring the potential graph).

Reply:
Should we be converting all of our dissimilarity functions to cython at some point?

Reply:
Yeah, this probably requires a bit more discussion. I can definitely see the benefit of using cython over numba though -- no need to go through the process of numbaizing the dissimilarity functions and we can probably cythonize the cluster dissimilarity function as well.
Though for most use cases cython is a bit slower than numba 😢